staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -232,6 +242,9 @@ func NewReflectorWithOptions(lw ListerWatcher, expectedType interface{}, store S
 		r.name = naming.GetNameFromCallsite(internalPackages...)
 	}
 
+	reflectorMetricsSuffix := atomic.AddInt64(&reflectorDisambiguator, 1)
+	r.metrics = newReflectorMetrics(fmt.Sprintf("reflector_"+options.Name+"_%d", reflectorMetricsSuffix))
```

Comment:
Err.. we're going to have cardinality issues with this approach. We've tried [this before](https://github.com/kubernetes/kubernetes/issues/52121#issuecomment-341001054), with horrible results.

Reply:
1.  For numberOfItemsInWatch/numberOfItemsInList, we can use gauge instead of histogram . It is not necessary to get quantiles 
2. For listDuration/watchDuration, we should reduce there buckets.
3. For other counter/gauge metrics, each will expose one value per reflector. I think it is necessary

We can try to reduce the resource consumption of metrics
- if name is empty, don't expose metrics
- reduce the length of the generated name. For example,  simplify `reflector_k8s.io/client-go/informers/coordination/v1/lease.go:59_6934` to `k8s.io/c/i/c/v1/lease.go:59_6934`

@logicalhan  WDYT?

Reply:
no i don't think that's going to help. I think we need to get rid of the name altogether.

Reply:
I think that metrics should contain meaningful information. As of now, the only available piece of information is the 'name', whether it's in the metrics name or in the labels.

For reflectors that have a specified 'name', we shouldn't use a Disambiguator to generate serial numbers. Users need to ensure that each reflector is unique, we can check it in `newReflectorMetrics`. Same reflector will has the `name` label. It will help to reduce number of cardinality.

For reflectors that generate a 'name' using the Callsite, we will refrain from generating metrics for now.

@logicalhan What do you think about this proposal?

Reply:
I don't think we can include the name, unless it's bounded, which I don't believe it is. Not generating metrics for reflectors using name generated from a callsite means that your reflector metrics, in aggregate, will not be accurate, which begs the question why would you implement a metric that you know isn't representing reality?

Reply:
We should definitely make sure the total is accurate, regardless of the labels we decide to break it down by.  Metrics for some reflectors but not others breaks this, so we shouldn't do that.

nit: I'm not a fan of this naming pattern. If we are going to add the reflector name as a label, we don't need to prefix/suffix it with other stuff.

Can we come up with a reasonable upper bound on the number of reflectors?  Is it bounded by the number of objects a controller is watching (which is usually < 10 ?)?  In terms of memory usage / memory leaks, if we are already storing a large amount of k8s objects in memory for each reflector, it is probably OK to have a few metrics that are broken down by the reflector name, assuming that represents a small fraction of the memory usage of the reflector overall.  @logicalhan is this the same as the issue you linked?  It looks like the issue might've been that it includes a resource version?  Is the resource version included here as well?

Reply:
Each reflector metric requires a unique identifier to prevent confusion between metrics. Within the reflector, it seems that only the "name" attribute has this potential.  @logicalhan Do you have any other suggestions, perhaps?

Reply:
https://github.com/kubernetes/kubernetes/pull/49919 add suffix to avoid duplicate metrics. I think it primarily addresses the issue of duplicate names generated from a callsite.

Currently, there doesn't seem to be an ideal solution to ensure the uniqueness of names generated from a callsite. Perhaps we might consider not generating metrics for it for the time being?

Reply:
At present, we are able to generate metrics for this type of reflector: https://github.com/kubernetes/kubernetes/blob/a823a842ecefa5391fe152c85ac99425120213b6/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L406

Reply:
For informer, maybe we can add `NewNamedInformer`

Reply:
>  Not generating metrics for reflectors using name generated from a callsite means that your reflector metrics, in aggregate, will not be accurate, which begs the question why would you implement a metric that you know isn't representing reality?

@logicalhan We can provide two types of metrics: one is the aggregate of all reflectors, and the other is individual metrics for named reflectors.
The aggregate metrics exclude the resource version.

Reply:
What is the cardinality of reflectors that we expect?

Reply:
> What is the cardinality of reflectors that we expect?

I think most operators will utilize reflector listwatch for fewer than 10 types of resources. These reflector metrics can be output by specifying their names.  The cardinality is less than 10. An example of this would be the kubelet.

For the APIServer and controller-manager, which watch all types of resources, it may be necessary to use parameters to determine which reflectors will expose metrics.

Reply:
I'm still not crazy about reflector names, mostly because we've been bitten by it before.

Reply:
@logicalhan If don't use name, it's difficult to distinguish. Do you have any suggestions?

Reply:
If don't use name, it's difficult to distinguish. Workqueue also uses name as the label https://github.com/kubernetes/kubernetes/blob/cc5362ebc17e1376fa79b510f7f354dbffe7f92e/staging/src/k8s.io/client-go/util/workqueue/metrics.go#L231

@logicalhan  Do you have any suggestions?

Reply:
@logicalhan I think reflector metrics would not become high cardinality if we only expose individual metrics for named reflector, it is similar to the situation with workqueue

Reply:
In kubernetes operator patterns, we use each reflector to watch one type of resource , i think it is bounded.  

We want to use these metrics to identify the root cause of reflectors(e.g. list delay, watch delay, etc.) i think it is difficult to solve this problem with logs